nav2_collision_monitor/src/collision_detector_node.cpp

```diff
@@ -0,0 +1,338 @@
+// Copyright (c) 2023 Samsung R&D Institute Russia
+// Copyright (c) 2023 Pixel Robotics GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_collision_monitor/collision_detector_node.hpp"
+
+#include <exception>
+#include <utility>
+#include <functional>
+
+#include "tf2_ros/create_timer_ros.h"
+
+#include "nav2_util/node_utils.hpp"
+
+using namespace std::chrono_literals;
+
+namespace nav2_collision_monitor
+{
+
+CollisionDetector::CollisionDetector(const rclcpp::NodeOptions & options)
+: nav2_util::LifecycleNode("collision_detector", "", options)
+{
+}
+
+CollisionDetector::~CollisionDetector()
+{
+  polygons_.clear();
+  sources_.clear();
+}
+
+nav2_util::CallbackReturn
+CollisionDetector::on_configure(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Configuring");
+
+  // Transform buffer and listener initialization
+  tf_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
+  auto timer_interface = std::make_shared<tf2_ros::CreateTimerROS>(
+    this->get_node_base_interface(),
+    this->get_node_timers_interface());
+  tf_buffer_->setCreateTimerInterface(timer_interface);
+  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
+
+  state_pub_ = this->create_publisher<nav2_msgs::msg::CollisionDetectorState>(
+    "collision_detector_state", rclcpp::SystemDefaultsQoS());
+
+  // Obtaining ROS parameters
+  if (!getParameters()) {
+    return nav2_util::CallbackReturn::FAILURE;
+  }
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+CollisionDetector::on_activate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Activating");
+
+  // Activating lifecycle publisher
+  state_pub_->on_activate();
+
+  // Activating polygons
+  for (std::shared_ptr<Polygon> polygon : polygons_) {
+    polygon->activate();
+  }
+
+  // Creating timer
+  timer_ = this->create_timer(
+    std::chrono::duration<double>{1.0 / frequency_},
+    std::bind(&CollisionDetector::process, this));
+
+  // Creating bond connection
+  createBond();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+CollisionDetector::on_deactivate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Deactivating");
+
+  // Deactivating lifecycle publishers
+  state_pub_->on_deactivate();
```

Comment:
`state_pub_` is better to deactivate after the `timer_` to be reset. Otherwise we might have (potential) multi-thread situation when timer callback will be called one more time after state publisher will be activated. This indeed won't appear in our case, as Nav2 is using single-threaded ROS-executor. But anyway and just for the case, it is more clear that everything is being stopped sequentially in the right order.